contracts/gol.cairo

```diff
@@ -0,0 +1,450 @@
+%lang starknet
+%builtins pedersen range_check bitwise
+
+from starkware.cairo.common.alloc import alloc
+from starkware.cairo.common.cairo_builtins import (HashBuiltin,
+    BitwiseBuiltin)
+from starkware.cairo.common.math import (assert_not_zero,
+    assert_le_felt, assert_not_equal, unsigned_div_rem, split_felt)
+from starkware.starknet.common.syscalls import get_caller_address
+from starkware.cairo.common.memcpy import memcpy
+from starkware.cairo.common.uint256 import Uint256
+
+from contracts.utils.life_rules import evaluate_rounds
+from contracts.utils.packing import (unpack_game, pack_game)
+from contracts.utils.constants import (DIM, CREATE_CREDIT_REQUIREMENT,
+    INFINITE_GAME_GENESIS, GIVE_LIFE_CREDIT_REQUIREMENT)
+
+from openzeppelin.upgrades.library import Proxy
+from openzeppelin.token.erc20.library import ERC20
+
+##### Description #####
+#
+# Creator is a version of GoL where user may create their own
+# starting point for the game. A player may participate only
+# once they have contributed to other games to evolve them.
+#
+# Infinite is a version of GoL where user may contribute to 
+# the game by evolving to next generation (and subsequently 
+# giving life to chosen dead cell).
+#
+#######################
+
+##### Storage #####
+# Records game history on chain.
+# Game id is to ensure game uniqueness.
+@storage_var
+func stored_game(
+        game_id : felt,
+        generation : felt,
+    ) -> (
+        owner_and_state : (felt, felt)
```

Comment:
Do we need to store owner here at all? Also using tuples is asking for trouble, please create a struct containing those two values.

Reply:
Let's see, after last changes I guess this in not essential to have on chain, since we'll have an ownership in indexer. I figured this is something that would be nice to have, to be able to verify the owner using contract. If you think, however, that it is not the case, I can remove this.

Reply:
Removing it means every operation is cheaper, so it might be a good idea.